File: RB101_RB109_small_problems/easy_4/convert_a_signed_number_to_a_string.rb
DIGITS = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']

def integer_to_string(num)
  strings = num.abs.digits.reverse.map {|num| DIGITS[num]}
  strings.join("")
end

def determine_sign(num)
  if num < 0 then "-"
  elsif num > 0 then "+"
  else ""
  end
end

def signed_integer_to_string(num)
  sign = determine_sign(num)
  sign + integer_to_string(num)
end

p signed_integer_to_string(4321) == '+4321'
p signed_integer_to_string(-123) == '-123'
p signed_integer_to_string(0) == '0'

#Further Exploration

def signed_integer_to_string(number)
  string_num = integer_to_string(number.abs)
  case number <=> 0
  when -1 then "-#{string_num}"
  when +1 then "+#{string_num}"
  else         string_num
  end
end

File: JS210_fundamentals/basics/convert_a_signed_number_to_a_string.js
function integerToString(num) {
  const BASE = 10;
  const DIGITS_TO_STRING = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'];
  let stringComponents = [];

  do {
    let digit = num % BASE;
    stringComponents.unshift(DIGITS_TO_STRING[digit]);
    num = Math.floor(num / BASE);
  } while (num > 0);

  return stringComponents.join('');
}

function signedIntegerToString(num) {
  let sign = '+';

  if (num < 0 || Object.is(num, -0)) {
    sign = '-';
    num *= -1;
  } else if (Object.is(num, 0)) {
    sign = '';
  }

  return sign + integerToString(num); 
}


console.log(signedIntegerToString(4321));      // "4321"
console.log(signedIntegerToString(0));         // "0"
console.log(signedIntegerToString(-123));      // "-123"
console.log(signedIntegerToString(-0) === "-0");

// LS Solution

// function signedIntegerToString(number) {
//   if (number < 0) {
//     return ('-' + integerToString(-number));
//   } else if (number > 0) {
//     return ('+' + integerToString(number));
//   } else {
//     return integerToString(number);
//   }
// }